Nama : Februwati Silalahi
Prodi : DIII-Teknologi Informasi
NIM: 42324047

Tugas
1. Ikuti semua tahapan dalam modul ini.

In [ ]:
import numpy as np
import pandas as pd
from sklearn.datasets import load_iris
# Load dataset Iris dari scikit-learn
iris = load_iris()


Pada sel In [1], kode berfungsi untuk memuat pustaka-pustaka dasar yang dibutuhkan dalam pengolahan data dan algoritma pembelajaran mesin. Pustaka numpy diimpor untuk menangani operasi matriks dan array numerik, sedangkan pandas diimpor untuk mengolah data dalam bentuk tabel atau DataFrame. Selain itu, fungsi load_iris diimpor dari modul sklearn.datasets untuk mengambil dataset bawaan Bunga Iris yang siap digunakan. Pada baris terakhir sel ini, fungsi load_iris() dipanggil dan disimpan dalam variabel iris

In [2]:
# Konversi data menjadi DataFrame Pandas
df = pd.DataFrame(iris.data, columns=iris.feature_names)
df["Species"] = iris.target_names[iris.target]
# Definisikan X (fitur) dan y (target/label)
X = df.drop(columns=["Species"])
y = df["Species"]
print(X)

     sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                  5.1               3.5                1.4               0.2
1                  4.9               3.0                1.4               0.2
2                  4.7               3.2                1.3               0.2
3                  4.6               3.1                1.5               0.2
4                  5.0               3.6                1.4               0.2
..                 ...               ...                ...               ...
145                6.7               3.0                5.2               2.3
146                6.3               2.5                5.0               1.9
147                6.5               3.0                5.2               2.0
148                6.2               3.4                5.4               2.3
149                5.9               3.0                5.1               1.8

[150 rows x 4 columns]


Pada sel In [2], data Iris yang telah dimuat diubah menjadi bentuk DataFrame Pandas dengan menentukan nama-nama kolom fiturnya menggunakan columns=iris.feature_names. Selanjutnya, dibuat kolom baru bernama Species yang memuat nama spesies asli bunga dengan memetakan indeks target ke nama kelasnya (iris.target_names[iris.target]). Setelah itu, variabel $X$ dipisahkan sebagai atribut/fitur dengan menghapus kolom Species, sedangkan variabel $y$ diambil khusus dari kolom Species sebagai label acuan. Luaran dari sel ini menampilkan isi variabel $X$ yang terdiri dari 150 baris data dan 4 kolom numerik, yaitu sepal length, sepal width, petal length, dan petal width

In [3]:
y

0         setosa
1         setosa
2         setosa
3         setosa
4         setosa
         ...    
145    virginica
146    virginica
147    virginica
148    virginica
149    virginica
Name: Species, Length: 150, dtype: str

Pada sel In [3], perintah y dijalankan untuk memeriksa isi dari variabel label target yang telah dipisahkan sebelumnya. Luaran Out [3] menampilkan daftar data target yang berupa teks/kategori nama spesies bunga, dimulai dari indeks 0 hingga 4 yang berisi spesies setosa, hingga indeks 145 hingga 149 yang berisi spesies virginica. Total data target ini berjumlah 150 baris dengan tipe data berupa objek atau teks.

In [4]:
from sklearn.preprocessing import LabelEncoder
lab = LabelEncoder()
y = lab.fit_transform(y)
y

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

Pada sel In [4], modul mengimpor kelas LabelEncoder dari sklearn.preprocessing untuk mengubah nilai label target yang berbentuk teks menjadi bentuk angka (numerik). Objek lab diinisialisasi, kemudian fungsi fit_transform(y) dijalankan untuk mengubah teks label target $y$ menjadi representasi angka secara otomatis. Luaran Out [4] memperlihatkan hasil konversi berupa array numerik, di mana angka 0 mewakili spesies setosa, angka 1 mewakili spesies versicolor, dan angka 2 mewakili spesies virginica

In [5]:
from sklearn.model_selection import train_test_split
X_train, X_test = train_test_split(X, test_size=0.2, random_state=0)
y_train, y_test = train_test_split(y, test_size=0.2, random_state=0)

Pada sel In [5], modul mengimpor fungsi train_test_split dari sklearn.model_selection untuk membagi dataset menjadi data latih (training) dan data uji (testing). Pembagian dilakukan untuk variabel fitur ($X$) dan variabel target ($y$) secara terpisah dengan proporsi data uji sebesar 20% (test_size=0.2), sehingga 80% sisa data digunakan untuk pelatihan model. Parameter random_state=0 ditambahkan agar pembagian data menghasilkan acakan yang konsisten dan dapat direplikasi setiap kali program dijalankan.

In [6]:
from sklearn.naive_bayes import GaussianNB
model = GaussianNB()
NB_model = model.fit(X_train, y_train)

ada sel In [6], modul memilih algoritma Gaussian Naïve Bayes karena seluruh fitur dalam dataset Iris bersifat numerik kontinu. Kelas GaussianNB diimpor dari sklearn.naive_bayes dan diinisialisasi ke dalam variabel model. Setelah itu, metode fit(X_train, y_train) dipanggil untuk melatih model Naïve Bayes menggunakan data latih yang telah dipisahkan sebelumnya, sehingga menghasilkan model terlatih bernama NB_model

In [7]:
result = NB_model.predict(X_test)
result

array([2, 1, 0, 2, 0, 2, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 0, 2, 1,
       0, 0, 2, 0, 0, 1, 1, 0])

Pada sel In [7], model yang sudah dilatih (NB_model) digunakan untuk memprediksi label kelas dari data uji X_test melalui fungsi predict(). Hasil prediksinya disimpan ke dalam variabel result. Luaran Out [7] menampilkan array angka berisi 30 nilai prediksi (berupa angka 0, 1, atau 2) yang merepresentasikan kelas hasil tebakan model untuk masing-masing sampel data uji.

In [8]:
from sklearn.metrics import accuracy_score
print('Accuracy of data testing: ', accuracy_score(y_test, result)*100)

Accuracy of data testing:  96.66666666666667


Pada sel In [8], modul mengimpor fungsi accuracy_score dari sklearn.metrics untuk mengukur tingkat keakuratan dari model yang dibuat. Fungsi ini membandingkan label sebenarnya dari data uji (y_test) dengan label hasil prediksi model (result), kemudian dikalikan 100 untuk mendapatkan nilai persentase. Luaran teks menampilkan bahwa nilai akurasi pengujian model pada data uji mencapai 96,67%.

In [9]:
#prediksi inputan data baru
#input_data = pd.DataFrame(np.array([[5.1, 3.5, 1.4, 0.2]]), columns=iris.feature_names)
input_data = [[5.1, 3.5, 1.4, 0.2]]
result = NB_model.predict(input_data)
prediction_label = lab.inverse_transform(result)
print('Hasil prediksi untuk input baru adalah:', prediction_label)

Hasil prediksi untuk input baru adalah: ['setosa']


c:\Users\ASUS\anaconda3\Lib\site-packages\sklearn\utils\validation.py:2827: UserWarning: X does not have valid feature names, but GaussianNB was fitted with feature names
  warnings.warn(


Pada sel In [9], modul mencoba melakukan prediksi terhadap satu sampel data baru dengan nilai fitur [[5.1, 3.5, 1.4, 0.2]] dalam bentuk list/array NumPy biasa. Nilai angka hasil prediksi kemudian dikembalikan ke bentuk nama spesies aslinya menggunakan lab.inverse_transform(result). Luaran mencetak hasil prediksi yaitu ['setosa'], namun sistem mengeluarkan pesan peringatan (UserWarning) karena model GaussianNB dilatih menggunakan DataFrame yang memiliki nama kolom fitur, sedangkan data input baru ini tidak memiliki nama kolom yang sesuai.

In [10]:
input_data = pd.DataFrame(np.array([[5.1, 3.5, 1.4, 0.2]]), columns=iris.feature_names)
result = NB_model.predict(input_data)
prediction_label = lab.inverse_transform(result)
print('Hasil prediksi untuk input baru adalah:', prediction_label)

Hasil prediksi untuk input baru adalah: ['setosa']


Pada sel In [10], modul memperbaiki masalah peringatan yang muncul pada langkah sebelumnya dengan mengubah input data baru menjadi format DataFrame Pandas yang menyertakan nama kolom (columns=iris.feature_names). Data baru yang berstruktur konsisten ini kemudian diprediksi kembali menggunakan fungsi predict() dan dikonversi ke label aslinya. Luaran yang dihasilkan tetap mencetak ['setosa'], namun kali ini berhasil dijalankan secara bersih tanpa adanya pesan peringatan dari sistem.   

Tugas 
2. Prediksi data baru masuk ke species apa: [7.3, 3.8, 4.6, 2.3]

In [11]:
input_data = pd.DataFrame(np.array([[7.3, 3.8, 4.6, 2.3]]), columns=iris.feature_names)
result = NB_model.predict(input_data)
prediction_label = lab.inverse_transform(result)
print('Hasil prediksi untuk input baru adalah:', prediction_label)

Hasil prediksi untuk input baru adalah: ['virginica']


Kode tersebut berfungsi untuk menguji data sampel baru menggunakan model Gaussian Naïve Bayes yang telah dilatih, di mana nilai fitur [7.3, 3.8, 4.6, 2.3] diubah menjadi DataFrame Pandas beserta nama kolomnya agar formatnya sesuai dengan data pelatihan dan terhindar dari pesan peringatan. Setelah itu, fungsi predict() menghitung kelas dengan probabilitas tertinggi, yang kemudian diubah dari format angka ke label teks aslinya menggunakan inverse_transform() sebelum ditampilkan ke layar. Output yang dihasilkan adalah ['virginica'] karena nilai lebar mahkota (petal width) sebesar 2.3 cm dan panjang kelopak (sepal length) sebesar 7.3 cm secara statistik merupakan ukuran dominan yang menjadi ciri khas utama spesies virginica pada dataset Iris.